godot-core/src/obj/raw.rs

```diff
@@ -0,0 +1,624 @@
+        sys::ptr_then(cast_object_ptr, |ptr| RawGd::from_obj_sys_weak(ptr))
+    }
+
+    pub(crate) fn as_ref_counted<R>(&self, apply: impl Fn(&mut engine::RefCounted) -> R) -> R {
+        debug_assert!(
+            self.is_instance_valid(),
+            "as_ref_counted() on freed instance; maybe forgot to increment reference count?"
+        );
+
+        let tmp = unsafe { self.ffi_cast::<engine::RefCounted>() };
+        let mut tmp = tmp.expect("object expected to inherit RefCounted");
+        let return_val =
+            <engine::RefCounted as GodotClass>::Declarer::scoped_mut(&mut tmp, |obj| apply(obj));
+
+        std::mem::forget(tmp); // no ownership transfer
+        return_val
+    }
+
+    pub(crate) fn as_object<R>(&self, apply: impl Fn(&mut engine::Object) -> R) -> R {
+        // Note: no validity check; this could be called by to_string(), which can be called on dead instances
+
+        let tmp = unsafe { self.ffi_cast::<engine::Object>() };
+        let mut tmp = tmp.expect("object expected to inherit Object; should never fail");
+        // let return_val = apply(tmp.inner_mut());
+        let return_val =
+            <engine::Object as GodotClass>::Declarer::scoped_mut(&mut tmp, |obj| apply(obj));
+
+        std::mem::forget(tmp); // no ownership transfer
+        return_val
+    }
+
+    // Target is always an engine class:
+    // * if T is an engine class => T
+    // * if T is a user class => T::Base
+    pub(super) fn as_target(
+        &self,
+    ) -> Option<&<<T as GodotClass>::Declarer as dom::Domain>::DerefTarget<T>> {
+        if self.is_null() {
+            return None;
+        }
+
+        // SAFETY:
+        // Every engine object is a struct like
+        //
+        // #[repr(C)]
+        // struct Node3D {
+        //     object_ptr: sys::GDExtensionObjectPtr,  // <- pointer
+        //     instance_id: InstanceId,                // <- non-zero u64
+        // }
+        //
+        // and `RawGd` looks like
+        //
+        // #[repr(C)]
+        // pub struct RawGd<T: GodotClass> {
+        //     pub(super) obj: *mut T,                 // <- pointer
+        //     cached_instance_id: Option<InstanceId>, // <- u64
+        // }
+        //
+        // So since self isn't null, that means `cached_instance_id` is not 0, and the two layouts are
+        // compatible.
+        let target = unsafe {
+            std::mem::transmute::<
+                &Self,
+                &<<T as GodotClass>::Declarer as dom::Domain>::DerefTarget<T>,
+            >(self)
+        };
+
+        Some(target)
+    }
+
+    // Target is always an engine class:
+    // * if T is an engine class => T
+    // * if T is a user class => T::Base
+    pub(super) fn as_target_mut(
+        &mut self,
+    ) -> Option<&mut <<T as GodotClass>::Declarer as dom::Domain>::DerefTarget<T>> {
+        if self.is_null() {
+            return None;
+        }
+
+        // SAFETY: see also [`as_target()`](Self::as_target)
```

Comment:
These won't generate documentation, so there are no links and a simple as_target (with or without backticks) is easier to read.